---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# panGenomeBreedr <img src="man/figures/logo.png" align="right" height="139" alt="" />

<!-- badges: start -->
<!-- badges: end -->

`panGenomeBreedr` (`panGB`) is conceptualized to be a unified, crop agnostic platform for pangenome-enabled breeding that follows standardized conventions for natural or casual variant analysis using pangenomes, marker design, and marker QC hypothesis testing (Figure 1). It seeks to simplify and enhance the use of pangenome resources in cultivar development.  

<figure style="text-align: center;">
  <img src="man/figures/workflow.png" width="800" alt="panGenomeBreedr marker design workflow">
  <figcaption><em>Fig. 1.</em> Conceptual workflow of the panGenomeBreedr (panGB) package for pangenome-enabled marker development. Using snpEff-annotated VCF files organized within a SQLite database, panGB enables querying for variants within candidate genes or user-defined genomic regions. It retrieves mutation annotations and predicted impacts from snpEff to identify putative causal variants (PCVs), which serve as the basis for designing functional trait-predictive markers. The package supports hypothesis-driven validation of these markers and also facilitates the design of additional marker types, including precision-introgression and background markers.</figcaption>
</figure>

In its current development version, panGB provides customizable R functions for **variant discovery from snpEff-annotated VCF files, KASP marker design, and marker validation** (Steps 1â€“3 in Fig. 1).  

To expand accessibility, `panGB` will include a user-friendly Shiny application, allowing non-R users to leverage its core features without requiring R programming experience.  

The SNP Viewer tool by LGC Genomics is limited to Windows platforms and lacks standardized conventions for visualizing positive controls in marker validation, making it difficult for users to conclusively assess marker performance. In contrast, `panGB` offers platform-independent tools for hypothesis testing, quality control (QC), and validation of KASP markers, addressing a key gap in existing visualization and validation workflows.  

Submit bug reports and feature suggestions, or track changes on the
[issues page](https://github.com/awkena/panGenomeBreedr/issues).

## Installation
First, ensure all existing packages are up to date.  

You can install the development version of `panGenomeBreedr` from [GitHub](https://github.com/awkena/panGenomeBreedr) with:

``` r
# Install panGenomeBreedr
if (!require("devtools")) install.packages("devtools")

devtools::install_github("awkena/panGenomeBreedr", upgrade = TRUE)
```

### Installing Bioconductor dependency packages
`panGB` depends on a list of Bioconductor packages that may not be installed automatically alongside `panGB`. To manually install these packages, use the code snippet below:

```r
# Install and load required Bioconductor packages
if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")

  BiocManager::install(c("Bsgenome",
                         "Biostrings",
                         "GenomicRanges",
                         "IRanges",
                         "msa"))
```

# Current Functionality of `panGB`
`panGB` currently provides functionality for the following key tasks:

1. **Variant discovery**  
   Identify variants within candidate genes or any user-defined genomic interval using *snpEff*-annotated VCF files.

2. **KASP marker design**  
   Generate allele-specific markers targeting either causal variants or any variant of interest.

3. **Marker validation and QC visualization**  
   Produce quality control plots and perform hypothesis-driven evaluations to assess marker reliability.

4. **Decision-support for trait introgression**  
   Guide marker-assisted backcrossing by profiling foreground, background, and precision-introgression markers to support selection decisions.

ðŸ‘‰ For a full tutorial and worked example, check out the
[panGenomeBreedr Workflow vignette](https://awkena.github.io/panGenomeBreedr/articles/panGenomeBreedr_Workflows.html).


## Troubleshooting

If the package does not run as expected, check the following:

-   Was the package properly installed?

-   Do you have the required dependencies installed?

-   Were any warnings or error messages returned during package
    installation?

-   Are all packages up to date before installing panGB?


# Authors and contributors

-   [Alexander Wireko Kena](https://www.github.com/awkena)

-   [Israel Tawiah Tetteh](https://github.com/Israel-Tetteh)

-   [Cruet Burgos](https://www.morrislab.org/people/clara-cruet-burgos)

-   [Linly Banda](https://www.biofortificationlab.org/people/linly-banda)

-   [Jacques Faye](https://sites.google.com/site/morrislaboratory/people/jacques-faye)

-   [Fanna Maina](https://www.morrislab.org/people/fanna-maina)

-   [Terry Felderhoff](https://www.agronomy.k-state.edu/about/people/faculty/felderhoff-terry/)

-   [Geoffrey Preston    Morris](https://www.morrislab.org/people/geoff-morris)



# License

[GNU GPLv3](https://choosealicense.com/licenses/gpl-3.0/)


# Support and Feedback

For support and submission of feedback, email the maintainer **Alexander Kena, PhD** at [alex.kena24@gmail.com](mailto:alex.kena24@gmail.com){.email}

