---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# panGenomeBreedr <img src="man/figures/logo.png" align="right" height="139" alt="" />

<!-- badges: start -->
<!-- badges: end -->

`panGenomeBreedr` (`panGB`) is conceptualized to be a unified, crop agnostic platform for pangenome-enabled breeding that follows standardized conventions for natural or casual variant analysis using pangenomes, marker design, and marker QC hypothesis testing (Figure 1). It seeks to simplify and enhance the use of pangenome resources in cultivar development.  

|<img src='man/figures/workflow.png' align="center" style="width: 800px;" />|
|:--:| 
| *Fig. 1. Conceptual workflow of the panGenomeBreedr (panGB) package for pangenome-enabled marker development. Using snpEff-annotated VCF files organized within a SQLite database, panGB enables querying for variants within candidate genes or user-defined genomic regions. It retrieves mutation annotations and predicted impacts from snpEff to identify putative causal variants (PCVs), which serve as the basis for designing functional trait-predictive markers. The package supports hypothesis-driven validation of these markers and also facilitates the design of additional marker types, including precision-introgression and background markers.* |.   

In its current development version, panGB provides customizable R functions for **variant discovery from snpEff-annotated VCF files, KASP marker design, and marker validation** (Steps 1â€“3 in Fig. 1).  

To expand accessibility, `panGB` will include a user-friendly Shiny application, allowing non-R users to leverage its core features without requiring R programming experience.  

The SNP Viewer tool by LGC Genomics is limited to Windows platforms and lacks standardized conventions for visualizing positive controls in marker validation, making it difficult for users to conclusively assess marker performance. In contrast, `panGB` offers platform-independent tools for hypothesis testing, quality control (QC), and validation of KASP markers, addressing a key gap in existing visualization and validation workflows.  

Submit bug reports and feature suggestions, or track changes on the
[issues page](https://github.com/awkena/panGenomeBreedr/issues).


# Table of contents

- [Requirements](#requirements)
- [Recommended packages](#recommended-packages)   
- [Installation](#installation)
- [Current Functionality of panGB](#current-functionality-of-pangb)
- [Usage](#usage) 
- [Examples](#examples) 
    - [Variant Discovery](#variant-discovery)
      - [Pangenome Data and Database Rationale](#pangenome-data-and-database-rationale)
      - [Recommended Schema for the SQLite Database](#recommended-schema-for-the-sqlite-database)
      - [Database Creation](#database-creation)
    - [KASP Marker Design](#kasp-marker-design)
    - [KASP Marker Validation](#kasp-marker-validation)
- [Other Breeder-Centered Functionalities in panGB](#other-breeder-centered-functionalities-in-pangb)
    - [Creating Heatmaps with `panGB`](#creating-heatmaps-with-pangb)
    - [Trait Introgression Hypothesis Testing](#trait-introgression-hypothesis-testing)
    - [Decision Support for Marker-Assisted Backcrossing in `panGB`](#decision-support-for-marker-assisted-backcrossing-in-pangb)
    - [Weighted RPP computation in panGB](#weighted-rpp-computation-in-pangb)
- [Troubleshooting](#troubleshooting)
- [Authors and contributors](#authors-and-contributors)
- [License](#license)
- [Support and Feedback](##support-and-feedback)


## Requirements
    
To run this package locally on a machine, the following R packages are
required:

-   [ggplot2](https://ggplot2.tidyverse.org): Elegant Graphics for Data Analysis.

-   [gridExtra](https://cran.r-project.org/web/packages/gridExtra/index.html):
    Miscellaneous Functions for "Grid" Graphics.
    
-   [utils](https://www.rdocumentation.org/packages/utils/versions/3.6.2):
    The R Utils Package.

-   [stats](https://www.rdocumentation.org/packages/stats/versions/3.6.2)

-   [reshape2](https://cran.r-project.org/web/packages/reshape2/index.html)
    
-   [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html)

-   [Biostrings](https://bioconductor.org/packages/release/bioc/html/Biostrings.html)

-   [GenomicRanges](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)

-   [IRanges](https://bioconductor.org/packages/release/bioc/html/IRanges.html)

-   [msa](https://bioconductor.org/packages/release/bioc/html/msa.html)


## Recommended packages

-   [Rtools](https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html): Needed for package development and installation from GitHub on Windows PCs.

-   [rmarkdown](https://CRAN.R-project.org/package=rmarkdown): When
    installed, display of the project's README.md will be rendered with R         Markdown.


## Installation
First, ensure all existing packages are up to date.  

You can install the development version of `panGenomeBreedr` from [GitHub](https://github.com/awkena/panGenomeBreedr) with:

``` r
# Install panGenomeBreedr
if (!require("devtools")) install.packages("devtools")

devtools::install_github("awkena/panGenomeBreedr", upgrade = TRUE)
```

### Installing Bioconductor dependency packages
`panGB` depends on a list of Bioconductor packages that may not be installed automatically alongside `panGB`. To manually install these packages, use the code snippet below:

```r
# Install and load required Bioconductor packages
if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")

  BiocManager::install(c("Bsgenome,
                         "Biostrings",
                         "GenomicRanges",
                         "IRanges",
                         "msa"))
```

# Current Functionality of `panGB`
`panGB` currently provides functionality for the following key tasks:

1. **Variant discovery**  
   Identify variants within candidate genes or any user-defined genomic interval using *snpEff*-annotated VCF files.

2. **KASP marker design**  
   Generate allele-specific markers targeting either causal variants or any variant of interest.

3. **Marker validation and QC visualization**  
   Produce quality control plots and perform hypothesis-driven evaluations to assess marker reliability.

4. **Decision-support for trait introgression**  
   Guide marker-assisted backcrossing by profiling foreground, background, and precision-introgression markers to support selection decisions.

# Usage
## Examples
Here, we provide examples on how to use `panGB` to design a KASP marker based on a causal variant, as well as marker validation for any KASP marker.

## Variant Discovery 
### Pangenome Data and Database Rationale
The examples used in this documentation are based on **sorghum pangenome resources** derived from whole-genome resequencing data of **1,676 sorghum lines**. Variant calling was performed using version **v5.1** of the **BTx623** reference genome. The resulting **SNP** and **INDEL** variants were functionally annotated using **snpEff**.

Direct querying of **snpEff-annotated VCF files** from R is often **computationally slow and inefficient**, especially with large datasets. To overcome this limitation, we built a **SQLite database** that stores the variants, annotations, and genotypes in normalized tables. This structure allows for **fast and flexible access** to relevant data, supporting workflows for trait-predictive marker discovery  

We **strongly recommend the creation of similar databases for other crops**. The SQLite format offers a compact, portable, and queryable representation of pangenome-derived variant data, significantly improving performance and reproducibility in genomic analysis pipelines.

A compressed format of the SQLite database for sorghum can be downloaded [here](https://drive.google.com/file/d/1L4S7_ZGeFyu_bA7rRsmTpf9V8__VLB-R/view?usp=sharing).

### Recommended Schema for the SQLite Database
The SQLite database contains the following three key tables:

#### `variants`

Stores core variant information extracted from the VCF.

| Column        | Description                        |
|---------------|------------------------------------|
| `variant_id`  | Unique variant identifier          |
| `chrom`       | Chromosome name                    |
| `pos`         | Genomic position (1-based)         |
| `ref`         | Reference allele                   |
| `alt`         | Alternate allele                   |
| `variant_type`| Type of variant (e.g., SNP, indel) |

#### `annotations`

Contains functional annotations from **snpEff**, typically including predicted effects, gene names, and functional categories.

| Column         | Description                               |
|----------------|-------------------------------------------|
| `variant_id`   | Foreign key linking to `variants`         |
| `gene_name`    | Sorghum gene ID (e.g., "Sobic.005G213600")|
| `effect`       | Type of effect (e.g., missense_variant)   |
| `impact`       | snpEff predicted impact (e.g., HIGH)      |
| `feature_type` | Type of annotated feature (e.g., exon)    |
| ...            | Additional snpEff annotation fields       |

#### `genotypes`

Stores genotype calls per sample for each variant.

| Column         | Description                            |
|----------------|----------------------------------------|
| `variant_id`   | Foreign key linking to `variants`      |
| `chrom`        | Chromosome name                        |
| `pos`          | Genomic position                       |
| `sample1`      | Genotype for Sample 1 (e.g., "A:G")     |
| `sample2`      | Genotype for Sample 2 (e.g., "A:A")     |
| ...            | Genotypes for other samples             |


### Database Creation
We generated the SQLite database a custom workflow that
1. Parses a multi-sample VCF file annotated by snpEff,
2. Extracts variant, annotation, and genotype data,
3. Writes the data into normalized relational tables (`variants`, `annotations`, `genotypes`).

A prebuilt mini example database (`mini_sorghum_variant_vcf.db.gz`) is included in the `extdata/` folder of the package.

### Query Variant Tables from SQLite Database

The `query_db()` function allows users to query specific tables within a panGenomeBreedr-formatted SQLite database for variants, annotations, or genotypes based on chromosome coordinates or candidate gene IDs.  

This function retrieves records from one of the following tables in the database:

- `variants`: Basic variant information (chromosome, position, REF/ALT alleles, etc.)
- `annotations`: Variant effect predictions (e.g., from snpEff)
- `genotypes`: Genotypic data across lines/samples plus the metadata of the variants.

Users can specify genomic coordinates (`chrom`, `start`, `end`) or a candidate gene name (`gene_name`) to extract relevant entries.

If used correctly, the `query_db()` function returns a data frame containing the filtered records from the selected table.

```{r, echo = TRUE}
library(panGenomeBreedr)

# This example uses the mini SQLite database included in the package.
# Define a temporary path and decompress the example database
path <- tempdir()
mini_db <- system.file("extdata", "mini_sorghum_variant_vcf.db.gz",
                       package = "panGenomeBreedr", mustWork = TRUE)
mini_db_path <- file.path(path, 'mini_sorghum_variant_vcf.db')
R.utils::gunzip(mini_db, destname = mini_db_path, remove = FALSE)

# Query VCF genotypes within the genomic range: Chr05:75104537-75106403
gt_region <- query_db(db_path = mini_db_path,
                      chrom = "Chr05",
                      start = 75104537,
                      end = 75106403,
                      table_name = "genotypes")

# Query snpEff annotations within a candidate gene
annota_region <- query_db(db_path = mini_db_path,
                          chrom = "Chr05",
                          start = 75104537,
                          end = 75106403,
                          table_name = "annotations",
                          gene_name = "Sobic.005G213600")

# Clean up temporary files
unlink(list.files(tempdir(), full.names = TRUE, recursive = TRUE), 
       recursive = TRUE)

knitr::kable(gt_region[1:5, 1:10], 
             caption = 'Table 1: Queried genotypes for varaints from the SQLite database.', 
             format = 'html', 
             padding = 0, 
             booktabs = TRUE, 
             escape = FALSE)

knitr::kable(annota_region[1:5,], 
             caption = 'Table 2: Queried annotations for variants from the SQLite database.', 
             format = 'html', 
             padding = 0, 
             booktabs = TRUE, 
             escape = FALSE)

```

The `query_db()` function has the following key arguments:
- `db_path`: *(character)* Path to the SQLite database file.
- `table_name`: *(character)* Name of the table to query (`"variants"`, `"annotations"`, or `"genotypes"`).
- `chrom`: *(character)* Chromosome name to filter by (e.g., `"Chr05"`).
- `start`, `end`: *(numeric)* Start and end genomic coordinates for the query region.
- `gene_name`: *(character)* Optional Sobic gene ID (e.g., `"Sobic.005G213600"`) â€” used when querying the `annotations` table to filter entries by gene.


### Filter Variants by Allele Frequency in a Genomic Region

The `query_by_af()` function allows users to extract variants from a SQLite database based on **alternate allele frequency thresholds** within a specified genomic region.

This is particularly useful for identifying **polymorphic sites** within candidate gene regions or windows of interest that meet desired minor allele frequency (MAF) thresholds for marker development.

An example usage for the `query_by_af()` function is shown in the code snippet below:

```{r, echo = TRUE}
library(panGenomeBreedr)

# Define temporary directory and decompress demo database
path <- tempdir()
mini_db <- system.file("extdata", "mini_sorghum_variant_vcf.db.gz",
                       package = "panGenomeBreedr", mustWork = TRUE)
mini_db_path <- file.path(path, 'mini_sorghum_variant_vcf.db')
R.utils::gunzip(mini_db, destname = mini_db_path, remove = FALSE)

# Filter variants with alt allele frequency between 1% and 99% in a defined region
filter_af <- query_by_af(db_path = mini_db_path,
                         min_af = 0.01,
                         max_af = 0.99,
                         chrom = "Chr05",
                         start = 75104537,
                         end = 75106403)

head(filter_af)

# Clean up temporary files
unlink(list.files(tempdir(), full.names = TRUE, recursive = TRUE), 
       recursive = TRUE)

knitr::kable(filter_af[1:5,], 
             caption = 'Table 3: Filtered variants from the SQLite database.', 
             format = 'html', 
             padding = 0, 
             booktabs = TRUE, 
             escape = FALSE)
```

The `query_by_af()` function has the following input parameters:

| Argument   | Type     | Description |
|-------------|----------|-------------|
| `db_path`   | `character` | Path to the SQLite database (`.db` file). |
| `min_af`    | `numeric`   | Minimum allele frequency (default: `0`). |
| `max_af`    | `numeric`   | Maximum allele frequency (default: `1`). |
| `chrom`     | `character` | Chromosome name (e.g., `"Chr05"`). |
| `start`     | `numeric`   | Start coordinate of the region. |
| `end`       | `numeric`   | End coordinate of the region. |


### Summarize SnpEff Annotation and Impact Distributions by Variant Type

The `query_ann_summary()` function provides a convenient way to summarize the distribution of **SnpEff annotations** and **impact categories** across variant types (e.g., SNPs, indels) within a defined genomic region.

This function enables users to quickly assess the types and functional implications of variants located within candidate genes or genomic intervals of interest.

```{r, echo = TRUE}
library(panGenomeBreedr)

# Prepare test database
path <- tempdir()
mini_db <- system.file("extdata", "mini_sorghum_variant_vcf.db.gz",
                       package = "panGenomeBreedr", mustWork = TRUE)
mini_db_path <- file.path(path, 'mini_sorghum_variant_vcf.db')
R.utils::gunzip(mini_db, destname = mini_db_path, remove = FALSE)

# Run annotation summary for region Chr05:75104537-75106403
ann_summary <- query_ann_summary(db_path = mini_db_path,
                                 chrom = "Chr05",
                                 start = 75104537,
                                 end = 75106403)

# View summaries
head(ann_summary$annotation_summary)
head(ann_summary$impact_summary)

# Clean up
unlink(list.files(tempdir(), full.names = TRUE, recursive = TRUE), 
       recursive = TRUE)

```

The `query_ann_summary()` function has the following input parameters:

| Argument            | Type        | Description |
|----------------------|-------------|-------------|
| `db_path`            | `character` | Path to the SQLite database. |
| `chrom`              | `character` | Chromosome name (e.g., `"Chr05"`). |
| `start`, `end`       | `numeric`   | Start and end coordinates of the target genomic region. |
| `annotations_table`  | `character` | Name of the table storing snpEff annotations (default: `"annotations"`). |
| `variants_table`     | `character` | Name of the table storing variant data (default: `"variants"`). |

The `query_ann_summary()` function returns a `list` with the following elements:
- `annotation_summary`: Data frame summarizing the count of each SnpEff annotation grouped by variant type.
- `impact_summary`: Data frame summarizing the count of each SnpEff impact level (e.g., HIGH, MODERATE) grouped by variant type.
- `variant_type_totals`: Total count of variants in the region grouped by variant type.

**The annotation summary shows that there are six (6) INDEL variants with a HIGH impact on protein function.** To see these variants, we need to use the `query_by_impact()` function, as shown below:

```{r, echo = TRUE}

# Prepare test database
path <- tempdir()
mini_db <- system.file("extdata", "mini_sorghum_variant_vcf.db.gz",
                       package = "panGenomeBreedr", mustWork = TRUE)
mini_db_path <- file.path(path, 'mini_sorghum_variant_vcf.db')
R.utils::gunzip(mini_db, destname = mini_db_path, remove = FALSE)

# Extract low impact variant for a region or gene
high_variants <- query_by_impact(db_path = mini_db_path,
                                impact_level = 'high',
                                chrom = "Chr05",
                                end = 75106403)

# View summaries
head(high_variants)

# Clean up
unlink(list.files(tempdir(), full.names = TRUE, recursive = TRUE), 
       recursive = TRUE)

```

## KASP Marker Design
The `kasp_marker_design()` function enables the design of **KASP (Kompetitive Allele Specific PCR)** markers from identified putative causal variants. It supports SNP, insertion, and deletion variants using VCF genotype data and a reference genome to generate Intertek-compatible marker information, including upstream and downstream polymorphic context.

This function automates the extraction of flanking sequences and polymorphic variants surrounding a focal variant and generates:

- Intertek-ready marker submission metadata
- DNA sequence alignment for visual inspection of marker context
- An optional publication-ready alignment plot in PDF format

The vcf file must contain the variant ID, Chromosome ID, Position, REF and ALT alleles, as well as the genotype data for samples, as shown in Table 1:

```{r, echo = FALSE}
head(gt_region)
```


```{r, echo = TRUE}
# Example to design a KASP marker on a HIGH impact Deletion variant
library(panGenomeBreedr)
path <- tempdir() # (default directory for saving alignment outputs)

# Path to import sorghum genome sequence for Chromosome 5
path1 <- "https://raw.githubusercontent.com/awkena/panGB/main/Chr05.fa.gz"

# KASP marker design for variant ID: INDEL_Chr05_75106156 in Table 1
lgs1 <- kasp_marker_design(gt_df = gt_region,
                           variant_id_col = 'variant_id',
                           chrom_col = 'chrom',
                           pos_col = 'pos',
                           ref_al_col = 'ref',
                           alt_al_col = 'alt',
                           genome_file = path1,
                           geno_start = 7,
                           marker_ID = "INDEL_Chr05_75106156",
                           chr = "Chr05",
                           plot_draw = TRUE,
                           plot_file = path,
                           region_name = "lgs1")


# View marker alignment output from temp folder
path3 <- file.path(path, list.files(path = path, "alignment_"))
system(paste0('open "', path3, '"')) # Open PDF file from R

on.exit(unlink(path)) # Clear the temp directory on exit

```

The `kasp_marker_design()` function has the following input parameters:

| Argument       | Type        | Description |
|------------------|-------------|-------------|
| `vcf_file`        | `character` | Path to VCF file containing variants. |
| `gt_df`           | `data.frame` or `matrix` |Pre-parsed genotype matrix with metadata and sample genotypes. |
| `variant_id_col`  | `character` | Column name for variant IDs. |
| `chrom_col`       | `character` | Column name for chromosome names. |
| `pos_col`         | `character` | Column name for variant positions. |
| `ref_al_col`      | `character` | Column name for reference alleles. |
| `alt_al_col`      | `character` | Column name for alternate alleles. |
| `geno_start`      | `integer`   | Column index in `gt_df` where genotype data starts. |
| `marker_ID`       | `character` | ID of the variant to design the marker for. |
| `chr`             | `character` | Chromosome name used to subset the reference genome. |
| `genome_file`     | `character` | Path to genome FASTA file. |
| `plot_draw`       | `logical`   | Whether to plot upstream/downstream alignment. |
| `plot_file`       | `character` | Output path to save the PDF alignment plot. |
| `region_name`     | `character` | Optional name for marker region. |
| `maf`             | `numeric`   | MAF filter for selecting flanking variants. |


The `kasp_marker_design()` function returns a `data.frame` with marker design metadata:

- `SNP_Name`: Variant ID
- `SNP`: Type of variant (SNP/INDEL)
- `Marker_Name`: Assigned name for the marker
- `Chromosome`: Chromosome name
- `Chromosome_Position`: Variant position
- `Sequence`: Intertek-style polymorphism sequence
- `ReferenceAllele`: Reference allele
- `AlternativeAllele`: Alternate allele

If `plot_draw = TRUE`, a **PDF plot** of sequence alignment will be saved to `plot_file`.


|<img src='man/figures/alignment.png' align="center" style="width: 700px;" />|
|:--:| 
| *Fig. 2. Alignment of the 100 bp upstream and downstream sequences to the reference genome used for KASP marker design.* |. 

The required sequence for submission to Intertek for the designed KASP marker is shown in Table 5.

```{r, echo = FALSE}
library(knitr)
knitr::kable(lgs1, 
             caption = 'Table 5: Intertek required sequence for a KASP marker.', 
             format = 'html', 
             padding = 0, 
             booktabs = TRUE, 
             escape = FALSE)
 
```



### KASP Marker Validation

The following example demonstrates how to use the customizable functions in `panGB` to perform hypothesis testing of allelic discrimination for KASP marker QC and validation.

`panGB` offers customizable functions for KASP marker validation through hypothesis testing. These functions allow users to easily perform the following tasks:  
-   Import raw or polished KASP genotyping results files (.csv) into R.   

-   Process imported data and assign FAM and HEX fluorescence colors for multiple     plates.  

-   Visualize marker QC using FAM and HEX fluorescence scores for each sample.  

-   Validate the effectiveness of trait-predictive or background markers using       positive controls.  

-   Visualize plate design and randomization.  


### Reading Raw KASP Full Results Files (.csv)
The `read_kasp_csv()` function allows users to import raw or polished KASP genotyping full results file (.csv) into R. The function requires the path of the raw file and the row tags for the different components of data in the raw file as arguments. 

For polished files, the user must extract the `Data` component of the full results file and save it as a csv file before import.

By default, a typical unedited raw KASP data file uses the following row tags for genotyping data: `Statistics`, `DNA`, `SNPs`, `Scaling`, `Data`.  

The raw file is imported as a list object in R. Thus, all components in the imported data can be extracted using the row tag ID as shown in the code snippet below:  

```{r}
# Import raw KASP genotyping file (.csv) using the read_kasp_csv() function
library(panGenomeBreedr)

# Set path to the directory where your data is located
# path1 <-  "inst/extdata/Genotyping_141.010_01.csv"
path1 <-  system.file("extdata", "Genotyping_141.010_01.csv",
                       package = "panGenomeBreedr",
                      mustWork = TRUE)

# Import raw data file
file1 <- read_kasp_csv(file = path1, 
                       row_tags = c("Statistics", "DNA", "SNPs", "Scaling", "Data"),
                       data_type = 'raw')

# Get KASP genotyping data for plotting
kasp_dat <- file1$Data
```

### Assigning colors and PCH symbols for KASP cluster plotting
The next step after importing data is to assign FAM and HEX fluorescence colors to samples based on their observed genotype calls. This step is accomplished using the `kasp_color()` function in `panGB` as shown in the code snippet below:  

```{r}
# Assign KASP fluorescence colors using the kasp_color() function
library(panGenomeBreedr)
# Create a subet variable called plates: masterplate x snpid
  kasp_dat$plates <- paste0(kasp_dat$MasterPlate, '_',
                                 kasp_dat$SNPID)
dat1 <- kasp_color(x = kasp_dat,
                    subset = 'plates',
                    sep = ':',
                    geno_call = 'Call',
                    uncallable = 'Uncallable',
                    unused = '?',
                    blank = 'NTC',
                   assign_cols = c(FAM = "blue", HEX = "gold" , 
                                   het = "forestgreen"))
```

The `kasp_color()` function requires the KASP genotype call file as a data frame and can do bulk processing if there are multiple master plates. The default values for the arguments in the `kasp_color()` function are based on KASP annotations. 

The `kasp_color()` function calls the `kasp_pch()` function to automatically add PCH plotting symbols that can equally be used to group genotypic clusters on the plot. 

When expected genotype calls are available for positive controls in KASP genotyping samples, we recommend the use of the PCH symbols for grouping observed genotypes instead of FAM and HEX colors.

The `kasp_color()` function expects that genotype calls are for diploid state with alleles separated by a symbol. By default KASP data are separated by `:` symbols.

The `kasp_color()` function returns a list object with the processed data for each master plate as the components.  

### Cluster plot
To test the hypothesis that the designed KASP marker can accurately discriminate between homozygotes and heterozygotes (allelic discrimination), a cluster plot needs to be generated.  

The `kasp_qc_ggplot()` and `kasp_qc_ggplot2()`functions in `panGB` can be used to make the cluster plots for each plate and KASP marker as shown below:   

```{r plate_05_qc_1, fig.cap = "Fig. 3. Cluster plot for Plate 5 using FAM and HEX colors for grouping observed genotypes."}
# KASP QC plot for Plate 05
library(panGenomeBreedr)
kasp_qc_ggplot2(x = dat1[5],
                    pdf = FALSE,
                    Group_id = NULL,
                    scale = TRUE,
                    expand_axis = 0.6,
                    alpha = 0.9,
                    legend.pos.x = 0.6,
                    legend.pos.y = 0.75)
```

```{r plate_05_qc_2, fig.cap = "Fig. 4. Cluster plot for Plate 5 with an overlay of predictions for positive controls."}
# KASP QC plot for Plate 05
library(panGenomeBreedr)
 kasp_qc_ggplot2(x = dat1[5],
                  pdf = FALSE,
                  Group_id = 'Group',
                  Group_unknown = '?',
                  scale = TRUE,
                  pred_cols = c('Blank' = 'black', 'False' = 'firebrick3',
                              'True' = 'cornflowerblue', 'Unverified' = 'beige'),
                  expand_axis = 0.6,
                  alpha = 0.9,
                  legend.pos.x = 0.6,
                  legend.pos.y = 0.75)
```
Color-blind-friendly color combinations are used to visualize verified genotype predictions (Figure 3).

In Figure 4, the three genotype classes are grouped based on plot PCH symbols using the FAM and HEX scores for observed genotype calls.

To simplify the verified prediction overlay for the expected genotypes for positive controls, all possible outcomes are divided into three categories (TRUE, FALSE, and UNVERIFIED) and color-coded to make it easier to visualize verified predictions.

BLUE (color code for the TRUE category) means genotype prediction matches the observed genotype call for the sample. 

RED (color code for the FALSE category) means genotype prediction does not match the observed genotype call for the sample. 

BEIGE (color code for the UNVERIFIED category) means three things: an expected genotype call could not be made before KASP genotyping, or an observed genotype call could not be made to verify the prediction. 

Users can set the `pdf = TRUE` argument to save plots as a PDF file in a directory outside R. The `kasp_qc_ggplot()` and `kasp_qc_ggplot2()`functions can generate cluster plots for multiple plates simultaneously.  

To visualize predictions for positive controls to validate KASP markers, the column name containing expected genotype calls must be provided and passed to the function using the `Group_id = 'Group'` argument as shown in the code snippets above. If this information is not available, set the argument `Group_id = NULL`. 

### Summary of Prediction Verification in Plates
The `pred_summary()` function produces a summary of predicted genotypes for positive controls in each reaction plate after verification (Table 3), as shown in the code snippet below:


```{r}
# Get prediction summary for all plates
library(panGenomeBreedr)
my_sum <- pred_summary(x = dat1,
                       snp_id = 'SNPID',
                       Group_id = 'Group',
                       Group_unknown = '?',
                       geno_call = 'Call',
                       rate_out = TRUE)
```

```{r, echo = FALSE}
library(knitr)
knitr::kable(my_sum$summ, caption = 'Table 3: Summary of verified prediction status for samples in plates', format = 'html', booktabs = TRUE)
 
```

The output of the `pred_summary()` function can be visualized as bar plots using the `pred_summary_plot()` function as shown in the code snippet below:

```{r barplot, fig.cap = "Fig. 5. Match/Mismatch rate of predictions for snp: snpSB00804."}
# Get prediction summary for snp:snpSB00804
library(panGenomeBreedr)
my_sum <- my_sum$summ
my_sum <- my_sum[my_sum$snp_id == 'snpSB00804',]

 pred_summary_plot(x = my_sum,
                    pdf = FALSE,
                    pred_cols = c('false' = 'firebrick3', 'true' = 'cornflowerblue',
                                  'unverified' = 'beige'),
                    alpha = 1,
                    text_size = 12,
                    width = 6,
                    height = 6,
                    angle = 45)


```

### Plot Plate Design
Users can visualize the observed genotype calls in a plate design format using the `plot_plate()` function as depicted in Figure 5, using the code snippet below:  

```{r plate_05_design, fig.cap = "Fig. 6. Observed genotype calls for samples in Plate 5 in a plate design format."}
plot_plate(dat1[5], pdf = FALSE)

```


# Other Breeder-Centered Functionalities in panGB
`panGB` provides additional functionalities to test hypotheses on the success of trait introgression pipelines and crosses. 

Users can easily generate heatmaps that compare the genetic background of parents to progenies to ascertain if a target locus was successfully introgressed or check for the hybridity of F1s. These plots also allow users to get a visual insight into the amount of parent germplasm recovered in progenies.  

To produce these plots, users must have either polymorphic low or mid-density marker data and a map file for the markers. **The map file must contain the marker IDs, their chromosome numbers and positions**.  

`panGB`can handle data from KASP, Agriplex and DArTag service providers.  


## Working with Agriplex Mid-Density Marker Data
Agriplex data is structurally different from KASP or DArTag data in terms of genotype call coding and formatting. Agriplex uses `' / '` as a separator for genotype calls for heterozygotes, and uses single nucleotides to represent homozygous SNP calls.

## Creating Heatmaps with panGB
To exemplify the steps for creating heatmap, we will use a mid-density marker data for three groups of near-isogenic lines (NILs) and their parents (Table 4). The NILs and their parents were genotyped using the Agriplex platform. Each NIL group was genotyped using 2421 markers.  

The imported data frame has the markers as columns and genotyped samples as rows. It comes with some meta data about the samples. Marker names are informative: chromosome number and position coordinates are embedded in the marker names (`Eg. S1_778962: chr = 1, pos = 779862`).  



```{r}

# Set path to the directory where your data is located
path1 <-  system.file("extdata", "agriplex_dat.csv",
                       package = "panGenomeBreedr",
                      mustWork = TRUE)

# Import raw Agriplex data file
geno <- read.csv(file = path1, header = TRUE, colClasses = c("character")) # genotype calls

library(knitr)
knitr::kable(geno[1:6, 1:10], caption = 'Table 4: Agriplex data format', format = 'html', booktabs = TRUE)


```

To create a heatmap that compares the genetic background of parents and NILs across all markers, we need to first process the raw Agriplex data into a numeric format. The panGB package has customizable data wrangling functions for KASP, Agriplex, and DArTag data.  

The `rm_mono()` function can be used to filter out all monomorphic loci from the data.  

Since our imported Agriplex data has informative SNP IDs, we can use the `parse_marker_ns()` function to generate a map file (Table 5) for the markers.    
The generated map file is then passed to the `proc_kasp()` function to order the SNP markers according to their chromosome numbers and positions. 

The `kasp_numeric()` function converts the output of the `proc_kasp()` function into a numeric format (Table 6).  The re-coding to numeric format is done as follows:  

* Homozygous for Parent 1 allele = 1. 
* Homozygous for Parent 2 allele = 0. 
* Heterozygous = 0.5. 
* Monomorphic loci = -1. 
* Loci with a suspected genotype error = -2. 
* Loci with at least one missing parental or any other genotype = -5. 



```{r}

# Parse snp ids to generate a map file
library(panGenomeBreedr)

# Data for stg5 NILs
stg5 <- geno[geno$Batch == 3, -c(1:6)] 
rownames(stg5) <- geno$Genotype[17:25]

# Remove monomorphic loci from data
stg5 <- rm_mono(stg5)

# Parse snp ids to generate a map file
snps <- colnames(stg5) # Get snp ids
map_file <- parse_marker_ns(x = snps, sep = '_', prefix = 'S')

# order markers in map file
map_file <- order_markers(x = map_file)

```

```{r, echo=FALSE}
library(knitr)
knitr::kable(map_file[1:8,], caption = 'Table 5: Map file for the imported Agriplex data.', format = 'html', booktabs = TRUE)

```



```{r}
# Process genotype data to re-order SNPs based on chromosome and positions
stg5 <- proc_kasp(x = stg5,
                  kasp_map = map_file,
                  map_snp_id = "snpid",
                  sample_id = "Genotype",
                  marker_start = 1,
                  chr = 'chr',
                  chr_pos = 'pos')

# Convert to numeric format for plotting
num_geno <- kasp_numeric(x = stg5,
                         rp_row = 1,
                         dp_row = 3,
                         sep = ' / ',
                         data_type = 'agriplex')

```


```{r, echo=FALSE}
library(knitr)
knitr::kable(num_geno[, 1:8], caption = 'Table 6: Agriplex data converted to a numeric format.', format = 'html', booktabs = TRUE)

```

All is now set to generate the heatmap (Figure 6) using the `cross_qc_ggplot()` function, as shown in the code snippet below:  


```{r heatmap1, fig.cap = "Fig. 6. A heatmap that compares the genetic background of parents and stg5 NIL progenies across all markers.", fig.width=12.5, fig.height=8.5}

# Get prediction summary for snp:snpSB00804
library(panGenomeBreedr)
# Create a heatmap that compares the parents to progenies
cross_qc_ggplot(x = num_geno,
                map_file = map_file,
                snp_ids = 'snpid',
                chr = 'chr',
                chr_pos = 'pos',
                parents = c("BTx623a", "BTx642a"),
                pdf = FALSE,
                filename = 'background_heatmap',
                legend_title = 'stg5_NILs',
                alpha = 0.9,
                text_size = 15)

```



The `cross_qc_ggplot()` function is a wrapper for functions in the `ggplot2` package.

Users must specify the IDs for the two parents using the `parents` argument. In the code snippet above, the recurrent parent is `BTx623` and the donor parent for the *stg5* locus is `BTx642`. 

The `group_sz` argument must be specified to plot the heatmap in batches of progenies to avoid cluttering the plot with many observations.  

Users can set the `pdf = TRUE` argument to save plots as a PDF file in a directory outside R.

## Trait Introgression Hypothesis Testing
To test the hypothesis that the *stg5* NIL development was effective, we can use the `cross_qc_annotate()`  function to generate a heatmap (Figure 7) with an annotation of the position of the *stg5* locus on Chr 1, as shown below:  

```{r heatmap2, fig.cap = "Fig. 7. Heatmap annotation of the stg5 locus on Chr 1.", fig.width=12.5, fig.height=6.5}

###########################################################################
# Subset data for the first 30 markers on Chr 1
stg5_ch1 <- num_geno[, map_file$chr == 1][,1:30] 

# Get the map file for subset data
stg5_ch1_map <- parse_marker_ns(colnames(stg5_ch1))

# Annotate a heatmap to show the stg5 locus on Chr 1
# The locus is between positions 0.98 - 1.8 Mbp on Chr 1
cross_qc_annotate(x = stg5_ch1,
                  map_file = stg5_ch1_map,
                  snp_ids = 'snpid',
                  chr = 'chr',
                  chr_pos = 'pos',
                  parents = c("BTx623a", "BTx642a"),
                  trait_pos = list(stg5 = c(start = .98e6, end = 1.8e6)),
                  text_scale_fct = 0.3,
                  pdf = FALSE,
                  legend_title = 'Stg5_NILs',
                  alpha = 0.9,
                  text_size = 15)

```


In the code snippet above, the numeric matrix of genotype calls and its associated map file are required. 

The recurrent and donor parents must be specified using the `parents` argument.

The `snp_ids, chr, and chr_pos` arguments can be used to specify the column names for marker IDs, chromosome number and positions in the attached map file.  
The `trait_pos` argument was used to specify the position of the target locus (*stg5*) on chromosome one. Users can specify the positions of multiple target loci as components of a list object for annotation.  

In Figure 7, the color intensity correlates positively with the marker density or coverage. Thus, areas with no color (white vertical gaps) depicts gaps in the marker coverage in the data.  



## Decision Support for Marker-Assisted Backcrossing in panGB
Users can use the `calc_rpp_bc()` function in `panGB` to calculate the proportion of recurrent parent background (RPP) fully recovered in backcross progenies. 

In the computation, partially regions are ignored, hence, heterozygous scores are not used.  

The output for he `calc_rpp_bc()` function can be passed to the `rpp_barplot()` function to visualize the computed RPP values for progenies as a bar plot. Users can specify an RPP threshold to easily identify lines that have RPP values above or equal to the defined RPP threshold on the bar plot.  

We can compute and visualize the observed RPP values for the *stg5* NILs across all polymorphic loci as shown in the code snippet below:  



```{r barplot_rpp1, fig.cap = "Fig. 8. Computed RPP values for the stg5 NILs.", fig.width=10.5, fig.height=6.5, }

# Calculate weighted RPP
rpp <- calc_rpp_bc(x = num_geno,
                   map_file = map_file,
                   map_chr = 'chr',
                   map_pos = 'pos',
                   map_snp_ids = 'snpid',
                   rp = 1,
                   rp_num_code = 1,
                   na_code = -5,
                   weighted = TRUE)

# Generate bar plot for RPP values
rpp_barplot(rpp_df = rpp,
            rpp_threshold = 0.93,
            text_size = 18,
            text_scale_fct = 0.1,
            alpha = 0.9,
            bar_width = 0.5,
            aspect_ratio = 0.5,
            pdf = FALSE)
```


The `calc_rpp_bc()` function in `panGB` provides two algorithms for computing the observed RPP values: weighted and unweighted RPP values. We recommend the use of the weighted algorithm to account for differences in the marker coverage across the genome. 

The algorithm for the weighted RPP values is explained below.


### Weighted RPP Computation in panGB
Let $w_i$ represent the weight for marker $i$, based on the relative distances to its adjacent markers. 

For a set of markers with positions $p_1, p_2, \ldots, p_n$, where $d_i = p_{i+1} - p_i$ represents the distance between adjacent markers, the weights can be calculated as follows:

1. **For the first marker** $i = 1$:  

   $$w_1 = \frac{d_1}{2 \sum_{i=1}^{n-1} d_i}$$

2. **For a middle marker** $1 < i < n$:  

   $$w_i = \frac{d_{i-1} + d_i}{2 \sum_{i=1}^{n-1} d_i}$$

3. **For the last marker** $i = n$:  

   $$w_n = \frac{d_{n-1}}{2 \sum_{i=1}^{n-1} d_i}$$

where:  

- $d_i$ is the distance between marker $i$ and marker $i+1$,
- $sum_{i=1}^{n-1} d_i$ is the total distance across all segments, used for normalization.  


Let $RPP$ represent the Recurrent Parent Proportion based on relative distance weighting. If $w_i$ is the weight for each marker $i$, and $m_i$ represents whether marker $i$ matches the recurrent parent $m_i = 1$ if it matches, $m_i = 0$ otherwise), then the weighted RPP is calculated as:  

$$RPP_{weighted} = \sum_{i=1}^n w_i\cdot m_i$$

The unweighted RPP is calculated without the use of the weights as follows:    

$$RPP_{unweighted} = \frac{\sum_{i=1}^n m_i} n$$

where:  

- $w_i$ is the weight of marker $i$, calculated based on the relative distance it covers,
- $m_i$ is the match indicator for marker $i$ (1 if matching the recurrent parent, 0 otherwise),
- $n$ is the total number of markers.

This formula provides the sum of the weighted contributions from each marker, representing the proportion of the recurrent parent genome in the individual.


## Troubleshooting

If the package does not run as expected, check the following:

-   Was the package properly installed?

-   Do you have the required dependencies installed?

-   Were any warnings or error messages returned during package
    installation?

-   Are all packages up to date before installing panGB?


# Authors and contributors

-   [Alexander Wireko Kena](https://www.github.com/awkena)

-   [Cruet Burgos](https://www.morrislab.org/people/clara-cruet-burgos)

-   [Linly Banda](https://www.biofortificationlab.org/people/linly-banda)

-   [Jacques Faye](https://sites.google.com/site/morrislaboratory/people/jacques-faye)

-   [Fanna Maina](https://www.morrislab.org/people/fanna-maina)

-   [Terry Felderhoff](https://www.agronomy.k-state.edu/about/people/faculty/felderhoff-terry/)

-   [Geoffrey Preston    Morris](https://www.morrislab.org/people/geoff-morris)



# License

[GNU GPLv3](https://choosealicense.com/licenses/gpl-3.0/)


# Support and Feedback

For support and submission of feedback, email the maintainer **Alexander Kena, PhD** at [alex.kena24@gmail.com](mailto:alex.kena24@gmail.com){.email}

