<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>panGenomeBreedr • panGenomeBreedr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="panGenomeBreedr">
<meta name="description" content="A unified platform for pangenome-enabled breeding that follows standardized conventions for natural or casual variant analysis using pangenomes, marker design, and marker QC visualizations.">
<meta property="og:description" content="A unified platform for pangenome-enabled breeding that follows standardized conventions for natural or casual variant analysis using pangenomes, marker design, and marker QC visualizations.">
<meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="pangenomebreedr-">panGenomeBreedr <a class="anchor" aria-label="anchor" href="#pangenomebreedr-"></a>
</h1>
</div>
<!-- badges: start -->
<!-- badges: end -->
<p><code>panGenomeBreedr</code> (<code>panGB</code>) is conceptualized to be a unified, crop agnostic platform for pangenome-enabled breeding that follows standardized conventions for natural or casual variant analysis using pangenomes, marker design, and marker QC hypothesis testing (Figure 1). It seeks to simplify and enhance the use of pangenome resources in cultivar development.</p>
<figure style="text-align: center;"><img src="reference/figures/workflow.png" width="800" alt="panGenomeBreedr marker design workflow"><figcaption><em>Fig. 1.</em> Conceptual workflow of the panGenomeBreedr (panGB) package for pangenome-enabled marker development. Using snpEff-annotated VCF files organized within a SQLite database, panGB enables querying for variants within candidate genes or user-defined genomic regions. It retrieves mutation annotations and predicted impacts from snpEff to identify putative causal variants (PCVs), which serve as the basis for designing functional trait-predictive markers. The package supports hypothesis-driven validation of these markers and also facilitates the design of additional marker types, including precision-introgression and background markers.
</figcaption></figure><p>In its current development version, panGB provides customizable R functions for <strong>variant discovery from snpEff-annotated VCF files, KASP marker design, and marker validation</strong> (Steps 1–3 in Fig. 1).</p>
<p>To expand accessibility, <code>panGB</code> will include a user-friendly Shiny application, allowing non-R users to leverage its core features without requiring R programming experience.</p>
<p>The SNP Viewer tool by LGC Genomics is limited to Windows platforms and lacks standardized conventions for visualizing positive controls in marker validation, making it difficult for users to conclusively assess marker performance. In contrast, <code>panGB</code> offers platform-independent tools for hypothesis testing, quality control (QC), and validation of KASP markers, addressing a key gap in existing visualization and validation workflows.</p>
<p>Submit bug reports and feature suggestions, or track changes on the <a href="https://github.com/awkena/panGenomeBreedr/issues" class="external-link">issues page</a>.</p>
</div>
<div class="section level1">
<h1 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h1>
<p>To use this package locally on a machine, the following R packages are required:</p>
<ul>
<li><p><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></p></li>
<li><p><a href="https://cran.r-project.org/web/packages/gridExtra/index.html" class="external-link">gridExtra</a></p></li>
<li><p><a href="https://www.rdocumentation.org/packages/utils/versions/3.6.2" class="external-link">utils</a></p></li>
<li><p><a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2" class="external-link">stats</a></p></li>
<li><p><a href="https://cran.r-project.org/web/packages/reshape2/index.html" class="external-link">reshape2</a></p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html" class="external-link">BSgenome</a></p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostrings</a></p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">GenomicRanges</a></p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html" class="external-link">IRanges</a></p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/msa.html" class="external-link">msa</a></p></li>
</ul>
</div>
<div class="section level1">
<h1 id="recommended-packages">Recommended packages<a class="anchor" aria-label="anchor" href="#recommended-packages"></a>
</h1>
<ul>
<li><p><a href="https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html" class="external-link">Rtools</a>: Needed for package development and installation from GitHub on Windows PCs.</p></li>
<li><p><a href="https://cran.r-project.org/web/packages/UpSetR/index.html" class="external-link">UpSetR</a>: Required for generating UpSet plots.</p></li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>First, ensure all existing packages are up to date.</p>
<p>You can install the development version of <code>panGenomeBreedr</code> from <a href="https://github.com/awkena/panGenomeBreedr" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install panGenomeBreedr</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"awkena/panGenomeBreedr"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="installing-bioconductor-dependency-packages">Installing Bioconductor dependency packages<a class="anchor" aria-label="anchor" href="#installing-bioconductor-dependency-packages"></a>
</h3>
<p><code>panGB</code> depends on a list of Bioconductor packages that may not be installed automatically alongside <code>panGB</code>. To manually install these packages, use the code snippet below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install and load required Bioconductor packages if not already installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define required Bioconductor packages</span></span>
<span><span class="va">bioc_packages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtracklayer"</span>, <span class="st">"Rsamtools"</span>, <span class="st">"msa"</span>, <span class="st">"IRanges"</span>,</span>
<span>                   <span class="st">"GenomicRanges"</span>, <span class="st">"BSgenome"</span>, <span class="st">"Biostrings"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install any missing Bioconductor packages</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">pkg</span> <span class="kw">in</span> <span class="va">bioc_packages</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="va">pkg</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">pkg</span>, ask <span class="op">=</span> <span class="cn">FALSE</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="current-functionality-of-pangb">Current Functionality of <code>panGB</code>
<a class="anchor" aria-label="anchor" href="#current-functionality-of-pangb"></a>
</h1>
<p><code>panGB</code> currently provides functionality for the following key tasks:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Variant discovery</strong><br>
Identify variants within candidate genes or any user-defined genomic interval using <em>snpEff</em>-annotated VCF files.</p></li>
<li><p><strong>KASP marker design</strong><br>
Generate allele-specific markers targeting either causal variants or any variant of interest.</p></li>
<li><p><strong>Marker validation and QC visualization</strong><br>
Produce quality control plots and perform hypothesis-driven evaluations to assess marker reliability.</p></li>
<li><p><strong>Decision-support for trait introgression</strong><br>
Guide marker-assisted backcrossing by profiling foreground, background, and precision-introgression markers to support selection decisions.</p></li>
</ol>
<p>👉 For a full tutorial and worked example, check out the <a href="https://awkena.github.io/panGenomeBreedr/articles/panGenomeBreedr_Workflows.html">panGenomeBreedr Workflow vignette</a>.</p>
<div class="section level2">
<h2 id="troubleshooting">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h2>
<p>If the package does not run as expected, check the following:</p>
<ul>
<li><p>Was the package properly installed?</p></li>
<li><p>Do you have the required dependencies installed?</p></li>
<li><p>Were any warnings or error messages returned during package installation?</p></li>
<li><p>Are all packages up to date before installing panGB?</p></li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="authors-and-contributors">Authors and contributors<a class="anchor" aria-label="anchor" href="#authors-and-contributors"></a>
</h1>
<ul>
<li><p><a href="https://www.github.com/awkena" class="external-link">Alexander Wireko Kena</a></p></li>
<li><p><a href="https://github.com/Israel-Tetteh" class="external-link">Israel Tawiah Tetteh</a></p></li>
<li><p><a href="https://www.morrislab.org/people/clara-cruet-burgos" class="external-link">Cruet Burgos</a></p></li>
<li><p><a href="https://www.biofortificationlab.org/people/linly-banda" class="external-link">Linly Banda</a></p></li>
<li><p><a href="https://sites.google.com/site/morrislaboratory/people/jacques-faye" class="external-link">Jacques Faye</a></p></li>
<li><p><a href="https://www.morrislab.org/people/fanna-maina" class="external-link">Fanna Maina</a></p></li>
<li><p><a href="https://www.agronomy.k-state.edu/about/people/faculty/felderhoff-terry/" class="external-link">Terry Felderhoff</a></p></li>
<li><p><a href="https://www.morrislab.org/people/geoff-morris" class="external-link">Geoffrey Preston Morris</a></p></li>
</ul>
</div>
<div class="section level1">
<h1 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h1>
<p><a href="https://choosealicense.com/licenses/gpl-3.0/" class="external-link">GNU GPLv3</a></p>
</div>
<div class="section level1">
<h1 id="support-and-feedback">Support and Feedback<a class="anchor" aria-label="anchor" href="#support-and-feedback"></a>
</h1>
<p>For support and submission of feedback, email the maintainer <strong>Alexander Kena, PhD</strong> at <a href="mailto:alex.kena24@gmail.com" class="email">alex.kena24@gmail.com</a></p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/awkena/panGenomeBreedr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/awkena/panGenomeBreedr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing panGenomeBreedr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alexander Wireko Kena <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3385-4945" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Israel Tawiah Tetteh <br><small class="roles"> Author </small>   </li>
<li>Clara Cruet Burgos <br><small class="roles"> Author </small>   </li>
<li>Geoffrey Preston Morris <br><small class="roles"> Contributor, funder, reviewer </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
