<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate the proportion of recurrent parent background (RPP) fully recovered in backcross progenies. — calc_rpp_bc • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate the proportion of recurrent parent background (RPP) fully recovered in backcross progenies. — calc_rpp_bc"><meta name="description" content="Calculate the proportion of recurrent parent background (RPP) fully recovered
in backcross progenies."><meta property="og:description" content="Calculate the proportion of recurrent parent background (RPP) fully recovered
in backcross progenies."><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate the proportion of recurrent parent background (RPP) fully recovered in backcross progenies.</h1>
      <small class="dont-index">Source: <a href="https://github.com/awkena/panGenomeBreedr/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
      <div class="d-none name"><code>calc_rpp_bc.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate the proportion of recurrent parent background (RPP) fully recovered
in backcross progenies.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_rpp_bc</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">map_file</span>,</span>
<span>  map_chr <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>  map_pos <span class="op">=</span> <span class="st">"pos"</span>,</span>
<span>  map_snp_ids <span class="op">=</span> <span class="st">"snpid"</span>,</span>
<span>  rp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rp_num_code <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  het_code <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  na_code <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  weighted <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A numeric matrix of marker genotypes for backcross progenies and recurrent
parent. Markers are columns and samples are rows.</p></dd>


<dt id="arg-map-file">map_file<a class="anchor" aria-label="anchor" href="#arg-map-file"></a></dt>
<dd><p>A data frame of map file consisting of marker IDs and their
chromosome numbers and positions as columns.</p></dd>


<dt id="arg-map-chr">map_chr<a class="anchor" aria-label="anchor" href="#arg-map-chr"></a></dt>
<dd><p>A character value indicating the column name for chromosome IDs
in <code>map_file</code>.</p></dd>


<dt id="arg-map-pos">map_pos<a class="anchor" aria-label="anchor" href="#arg-map-pos"></a></dt>
<dd><p>A character value indicating the column name for chromosome
positions in map file.</p></dd>


<dt id="arg-map-snp-ids">map_snp_ids<a class="anchor" aria-label="anchor" href="#arg-map-snp-ids"></a></dt>
<dd><p>A character value indicating the column name for SNP IDs
in <code>map_file</code>.</p></dd>


<dt id="arg-rp">rp<a class="anchor" aria-label="anchor" href="#arg-rp"></a></dt>
<dd><p>An integer or character value indicating the row index or sample ID
for the recurrent parent. if `NULL` the sample on the first row of <code>x</code> is
used as the recurrent parent.</p></dd>


<dt id="arg-rp-num-code">rp_num_code<a class="anchor" aria-label="anchor" href="#arg-rp-num-code"></a></dt>
<dd><p>A numeric value indicating the coding for the recurrent parent
marker background in <code>x</code>.</p></dd>


<dt id="arg-het-code">het_code<a class="anchor" aria-label="anchor" href="#arg-het-code"></a></dt>
<dd><p>A numeric value indicating the coding for heterozygous
marker background in <code>x</code>.</p></dd>


<dt id="arg-na-code">na_code<a class="anchor" aria-label="anchor" href="#arg-na-code"></a></dt>
<dd><p>A value indicating missing data in <code>x</code>.</p></dd>


<dt id="arg-weighted">weighted<a class="anchor" aria-label="anchor" href="#arg-weighted"></a></dt>
<dd><p>A logical value indicating whether RPP values should be weighted
or not.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame object comprising the RPP per chromosome and total RPP
of sample IDs.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The weighted RPP is computed as the sum of the product of relative distance
weighting of markers and genotype scores that match the recurrent parent.
Marker weights are obtained as half the normalized ratio of the physical
distance between marker `i` and marker `i+1` to the total distance for all
ordered markers on a chromosome.</p>
<p>The computation excludes heterozygous loci and only considers chromosome regions
fully recovered as recurrent parent genetic background.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://awkena.github.io/panGenomeBreedr">panGenomeBreedr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Create a numeric matrix of genotype scores for 10 markers and 5 samples</span></span></span>
<span class="r-in"><span><span class="va">num_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                    <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                    <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span> <span class="op">)</span>,</span></span>
<span class="r-in"><span>                  byrow <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">num_dat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'rp'</span>, <span class="st">'dp'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bc1_'</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">num_dat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'S1'</span>, <span class="st">'_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">10000</span>, len <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                            <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'S2'</span>, <span class="st">'_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">10000</span>, len <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get map file by parsing SNP IDs</span></span></span>
<span class="r-in"><span><span class="va">map_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="parse_marker_ns.html">parse_marker_ns</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">num_dat</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate weighted RPP</span></span></span>
<span class="r-in"><span><span class="va">rpp</span> <span class="op">&lt;-</span> <span class="fu">calc_rpp_bc</span><span class="op">(</span>x <span class="op">=</span> <span class="va">num_dat</span>,</span></span>
<span class="r-in"><span>                      map_file <span class="op">=</span> <span class="va">map_file</span>,</span></span>
<span class="r-in"><span>                      map_chr <span class="op">=</span> <span class="st">'chr'</span>,</span></span>
<span class="r-in"><span>                      map_pos <span class="op">=</span> <span class="st">'pos'</span>,</span></span>
<span class="r-in"><span>                      map_snp_ids <span class="op">=</span> <span class="st">'snpid'</span>,</span></span>
<span class="r-in"><span>                      rp <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                      rp_num_code <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                      weighted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

