<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Identify lines that possess favorable alleles for target loci using trait predictive markers. — foreground_select • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify lines that possess favorable alleles for target loci using trait predictive markers. — foreground_select"><meta name="description" content="Identify lines that possess favorable alleles for target loci using trait
predictive markers."><meta property="og:description" content="Identify lines that possess favorable alleles for target loci using trait
predictive markers."><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identify lines that possess favorable alleles for target loci using trait predictive markers.</h1>
      <small class="dont-index">Source: <a href="https://github.com/awkena/panGenomeBreedr/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
      <div class="d-none name"><code>foreground_select.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Identify lines that possess favorable alleles for target loci using trait
predictive markers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">foreground_select</span><span class="op">(</span></span>
<span>  <span class="va">geno_data</span>,</span>
<span>  <span class="va">fore_marker_info</span>,</span>
<span>  <span class="va">fore_marker_col</span>,</span>
<span>  <span class="va">fav_allele_col</span>,</span>
<span>  <span class="va">alt_allele_col</span>,</span>
<span>  select_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"homo"</span>, <span class="st">"hetero"</span>, <span class="st">"both"</span><span class="op">)</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">":"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-geno-data">geno_data<a class="anchor" aria-label="anchor" href="#arg-geno-data"></a></dt>
<dd><p>A data frame or matrix of marker genotype data with lines as
rows and markers as columns.</p></dd>


<dt id="arg-fore-marker-info">fore_marker_info<a class="anchor" aria-label="anchor" href="#arg-fore-marker-info"></a></dt>
<dd><p>A data frame of meta data for the trait-predictive markers
consisting of marker names, favorable and alternate alleles as variables.</p></dd>


<dt id="arg-fore-marker-col">fore_marker_col<a class="anchor" aria-label="anchor" href="#arg-fore-marker-col"></a></dt>
<dd><p>A character value specifying the column name for trait-predictive
markers in <code>fore_marker_info</code>.</p></dd>


<dt id="arg-fav-allele-col">fav_allele_col<a class="anchor" aria-label="anchor" href="#arg-fav-allele-col"></a></dt>
<dd><p>A character value specifying the column name for favorable
allele in <code>fore_marker_info</code>.</p></dd>


<dt id="arg-alt-allele-col">alt_allele_col<a class="anchor" aria-label="anchor" href="#arg-alt-allele-col"></a></dt>
<dd><p>A character value specifying the column name for alternate
allele in <code>fore_marker_info</code>.</p></dd>


<dt id="arg-select-type">select_type<a class="anchor" aria-label="anchor" href="#arg-select-type"></a></dt>
<dd><p>A character value of one of three options: `homo` to select
lines that are homozygous for the favorable allele; `hetero` to select lines
that are heterozygous; and `both` to select both favorable homozygotes and
heterozygotes.</p></dd>


<dt id="arg-sep">sep<a class="anchor" aria-label="anchor" href="#arg-sep"></a></dt>
<dd><p>A character value representing the separator used for genotype calling.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A binary data frame conversion of the original marker genotype data
for presence(1s) and absence(0s) of favorable alleles of target loci.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://awkena.github.io/panGenomeBreedr">panGenomeBreedr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Marker genotype data</span></span></span>
<span class="r-in"><span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>SNP1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A:A"</span>, <span class="st">"A:G"</span>, <span class="st">"G:G"</span>, <span class="st">"A:A"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   SNP2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C:C"</span>, <span class="st">"C:T"</span>, <span class="st">"T:T"</span>, <span class="st">"C:T"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   SNP3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G:G"</span>, <span class="st">"G:G"</span>, <span class="st">"A:G"</span>, <span class="st">"A:A"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Line1"</span>, <span class="st">"Line2"</span>, <span class="st">"Line3"</span>, <span class="st">"Line4"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Trait predictive markers meta data</span></span></span>
<span class="r-in"><span><span class="va">marker_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>qtl_markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          fav_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'C'</span>, <span class="st">'G'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          alt_alleles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'G'</span>, <span class="st">'T'</span>, <span class="st">'A'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Select lines where genotype is either homozygous for favorable alleles at all loci</span></span></span>
<span class="r-in"><span><span class="fu">foreground_select</span><span class="op">(</span>geno_data <span class="op">=</span> <span class="va">geno</span>,</span></span>
<span class="r-in"><span>                  fore_marker_info <span class="op">=</span> <span class="va">marker_info</span>,</span></span>
<span class="r-in"><span>                  fore_marker_col <span class="op">=</span> <span class="st">'qtl_markers'</span>,</span></span>
<span class="r-in"><span>                  fav_allele_col <span class="op">=</span> <span class="st">'fav_alleles'</span>,</span></span>
<span class="r-in"><span>                  alt_allele_col <span class="op">=</span> <span class="st">'alt_alleles'</span>,</span></span>
<span class="r-in"><span>                  select_type <span class="op">=</span> <span class="st">"homo"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       SNP1 SNP2 SNP3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Line1    1    1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Line2    0    0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Line3    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Line4    1    0    0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

