<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="variant-discovery-kasp-marker-design">Variant Discovery &amp; KASP Marker Design<a class="anchor" aria-label="anchor" href="#variant-discovery-kasp-marker-design"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="query_db.html">query_db()</a></code>

        </dt>
        <dd>Query any table in your SQLite database using chromosome and a genomic position range.</dd>
      </dl><dl><dt>

          <code><a href="query_by_af.html">query_by_af()</a></code>

        </dt>
        <dd>Extract variants based on minimum and maximum allele frequencies within a defined region in a SQLite database.</dd>
      </dl><dl><dt>

          <code><a href="query_ann_summary.html">query_ann_summary()</a></code>

        </dt>
        <dd>Query the annotations table within a specified genomic region and summarize the distribution of SnpEff annotations and impact categories by variant type.</dd>
      </dl><dl><dt>

          <code><a href="kasp_marker_design.html">kasp_marker_design()</a></code>

        </dt>
        <dd>Design KASP markers based on causal variants.</dd>
      </dl><dl><dt>

          <code><a href="list_sqlite_tables.html">list_sqlite_tables()</a></code>

        </dt>
        <dd>List all tables in the SQLite database.</dd>
      </dl><dl><dt>

          <code><a href="variant_stats.html">variant_stats()</a></code>

        </dt>
        <dd>Get variants statistics stored in SQLite database</dd>
      </dl><dl><dt>

          <code><a href="variant_impact_summary.html">variant_impact_summary()</a></code>

        </dt>
        <dd>Get variants statistics stored in SQLite database based on mutation impact.</dd>
      </dl><dl><dt>

          <code><a href="summarize_sqlite_tables.html">summarize_sqlite_tables()</a></code>

        </dt>
        <dd>Name and row count for each table in SQLite database.</dd>
      </dl><dl><dt>

          <code><a href="list_table_columns.html">list_table_columns()</a></code>

        </dt>
        <dd>Check the column names and types for any table in a SQLite database.</dd>
      </dl><dl><dt>

          <code><a href="gene_coord_gff.html">gene_coord_gff()</a></code>

        </dt>
        <dd>Get the genomic range of a candidate gene using the Sobic ID from a gff file.</dd>
      </dl><dl><dt>

          <code><a href="query_by_impact.html">query_by_impact()</a></code>

        </dt>
        <dd>Extract variants from annotation table based on impact type: LOW, MODERATE, HIGH, MODIFIER.</dd>
      </dl><dl><dt>

          <code><a href="calc_af.html">calc_af()</a></code>

        </dt>
        <dd>Compute allele frequencies for a VCF genotype matrix (variant x samples). Chromosome and position may be included in the data.</dd>
      </dl><dl><dt>

          <code><a href="filter_by_af.html">filter_by_af()</a></code>

        </dt>
        <dd>Filter extracted variants based on alternate allele frequency.</dd>
      </dl><dl><dt>

          <code><a href="query_genotypes.html">query_genotypes()</a></code>

        </dt>
        <dd>Query genotypes for one or more variant IDs from a wide-format genotype table.</dd>
      </dl><dl><dt>

          <code><a href="count_variant_types.html">count_variant_types()</a></code>

        </dt>
        <dd>Count the number of variant types in the SQLite database.</dd>
      </dl><dl><dt>

          <code><a href="extract_variant.html">extract_variant()</a></code>

        </dt>
        <dd>Extract putative causal variants within a candidate gene from a tabix-indexed snpEff annotated VCF file.</dd>
      </dl><dl><dt>

          <code><a href="get_google_id.html">get_google_id()</a></code>

        </dt>
        <dd>Get the folder or file ID from a Google Drive shareable link.</dd>
      </dl><dl><dt>

          <code><a href="folder_download_gd.html">folder_download_gd()</a></code>

        </dt>
        <dd>Download files in a shared Google Drive folder without restrictions.</dd>
      </dl></div><div class="section level2">
      <h2 id="kasp-marker-validation">KASP Marker Validation<a class="anchor" aria-label="anchor" href="#kasp-marker-validation"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="read_kasp_csv.html">read_kasp_csv()</a></code>

        </dt>
        <dd>Read raw KASP results file (csv format) with one or multiple plates.</dd>
      </dl><dl><dt>

          <code><a href="get_alleles.html">get_alleles()</a></code>

        </dt>
        <dd>Get SNP or InDel alleles and possible genotypes from genotype calls in KASP data.</dd>
      </dl><dl><dt>

          <code><a href="kasp_pch.html">kasp_pch()</a></code>

        </dt>
        <dd>Generate pch characters for cluster plots of KASP genotype calls.</dd>
      </dl><dl><dt>

          <code><a href="kasp_color.html">kasp_color()</a></code>

        </dt>
        <dd>Color-code KASP genotype calls based on LGC genomics colors for HEX and FAM.</dd>
      </dl><dl><dt>

          <code><a href="scale_axis.html">scale_axis()</a></code>

        </dt>
        <dd>Normalize FAM and HEX fluorescence values between 0 and 1</dd>
      </dl><dl><dt>

          <code><a href="pred_status.html">pred_status()</a></code>

        </dt>
        <dd>Generate the prediction status of positive controls in a KASP assay, if present.</dd>
      </dl><dl><dt>

          <code><a href="pred_summary.html">pred_summary()</a></code>

        </dt>
        <dd>Generate summary of prediction for positive controls in KASP genotype data, if present</dd>
      </dl><dl><dt>

          <code><a href="kasp_qc_ggplot.html">kasp_qc_ggplot()</a></code>

        </dt>
        <dd>Make KASP marker genotyping QC plot.</dd>
      </dl><dl><dt>

          <code><a href="kasp_qc_ggplot2.html">kasp_qc_ggplot2()</a></code>

        </dt>
        <dd>Make KASP marker genotyping QC plot overlaid with predicitons.</dd>
      </dl><dl><dt>

          <code><a href="plot_plate.html">plot_plate()</a></code>

        </dt>
        <dd>Plot kasp genotyping plate layout.</dd>
      </dl><dl><dt>

          <code><a href="nsamples_plate.html">nsamples_plate()</a></code>

        </dt>
        <dd>Get a summary of the number of samples per 96-well plate in a multi-plate KASP assay.</dd>
      </dl><dl><dt>

          <code><a href="kasp_reshape_wide.html">kasp_reshape_wide()</a></code>

        </dt>
        <dd>Reshape KASP data to wide format for same samples genotyped with multiple KASP markers.</dd>
      </dl><dl><dt>

          <code><a href="proc_kasp.html">proc_kasp()</a></code>

        </dt>
        <dd>Process reshaped KASP genotype data for heatmap plotting</dd>
      </dl><dl><dt>

          <code><a href="geno_error.html">geno_error()</a></code>

        </dt>
        <dd>Identify SNP loci with potential genotype call errors.</dd>
      </dl><dl><dt>

          <code><a href="kasp_numeric.html">kasp_numeric()</a></code>

        </dt>
        <dd>Convert processed KASP data to numeric genotypes</dd>
      </dl><dl><dt>

          <code><a href="pred_summary_plot.html">pred_summary_plot()</a></code>

        </dt>
        <dd>Create decision support bar plots of match vs. mismatch rates of KASP markers that had predictions for positive controls.</dd>
      </dl><dl><dt>

          <code><a href="gg_dat.html">gg_dat()</a></code>

        </dt>
        <dd>Convert data matrix for genotypes to a long format data frame.</dd>
      </dl></div><div class="section level2">
      <h2 id="decision-support-for-mabc-and-trait-introgression">Decision Support for MABC and Trait Introgression<a class="anchor" aria-label="anchor" href="#decision-support-for-mabc-and-trait-introgression"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="parse_marker_ns.html">parse_marker_ns()</a></code>

        </dt>
        <dd>Parse marker names of Hapmap format with a common pattern containing chromosome numbers and positions into a map file.</dd>
      </dl><dl><dt>

          <code><a href="cross_qc_ggplot.html">cross_qc_ggplot()</a></code>

        </dt>
        <dd>Create a heatmap to visualize and compare the genetic genetic backgrounds of genotypes/lines.</dd>
      </dl><dl><dt>

          <code><a href="rm_mono.html">rm_mono()</a></code>

        </dt>
        <dd>Remove or filter out monomorphic loci from a data matrix or frame.</dd>
      </dl><dl><dt>

          <code><a href="calc_rpp_bc.html">calc_rpp_bc()</a></code>

        </dt>
        <dd>Calculate the proportion of recurrent parent background (RPP) fully recovered in backcross progenies.</dd>
      </dl><dl><dt>

          <code><a href="calc_rpp_exp.html">calc_rpp_exp()</a></code>

        </dt>
        <dd>Compute theoretical RPP values for any specified backcross generation.</dd>
      </dl><dl><dt>

          <code><a href="rpp_barplot.html">rpp_barplot()</a></code>

        </dt>
        <dd>Visualize computed RPP values for BC progenies as a bar plot.</dd>
      </dl><dl><dt>

          <code><a href="cross_qc_annotate.html">cross_qc_annotate()</a></code>

        </dt>
        <dd>Annotate a heatmap to show introgressed loci positions.</dd>
      </dl><dl><dt>

          <code><a href="cross_qc_heatmap.html">cross_qc_heatmap()</a></code>

        </dt>
        <dd>Create a heatmap to visualize and compare the genetic genetic backgrounds of genotypes/lines with or without annotation for introgressed loci.</dd>
      </dl><dl><dt>

          <code><a href="sim_snp_dat.html">sim_snp_dat()</a></code>

        </dt>
        <dd>Simulate raw SNP loci for any chromosome with or without LD.</dd>
      </dl><dl><dt>

          <code><a href="order_markers.html">order_markers()</a></code>

        </dt>
        <dd>Order marker IDs based on their chromosome numbers and positions in ascending order.</dd>
      </dl><dl><dt>

          <code><a href="hapmap_ns_fmt.html">hapmap_ns_fmt()</a></code>

        </dt>
        <dd>Format marker names to comply with the Hapmap convention.</dd>
      </dl><dl><dt>

          <code><a href="find_unexp_homs.html">find_unexp_homs()</a></code>

        </dt>
        <dd>Find loci with unexpected homozygous genotype calls for artificial heterozygotes.</dd>
      </dl><dl><dt>

          <code><a href="find_indels.html">find_indels()</a></code>

        </dt>
        <dd>Identify and subset InDel markers from a marker panel.</dd>
      </dl><dl><dt>

          <code><a href="parent_missing.html">parent_missing()</a></code>

        </dt>
        <dd>Identify and subset loci with any parent missing genotype.</dd>
      </dl><dl><dt>

          <code><a href="parent_het.html">parent_het()</a></code>

        </dt>
        <dd>Identify and subset loci with any heterozygous parent genotype.</dd>
      </dl><dl><dt>

          <code><a href="parent_poly.html">parent_poly()</a></code>

        </dt>
        <dd>Select polymorphic loci between two parents in a marker panel.</dd>
      </dl><dl><dt>

          <code><a href="foreground_select.html">foreground_select()</a></code>

        </dt>
        <dd>Identify lines that possess favorable alleles for target loci using trait predictive markers.</dd>
      </dl><dl><dt>

          <code><a href="find_lines.html">find_lines()</a></code>

        </dt>
        <dd>Extracts lines that have a combination of favorable alleles across target loci.</dd>
      </dl></div><div class="section level2">
      <h2 id="example-datasets">Example Datasets<a class="anchor" aria-label="anchor" href="#example-datasets"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="beta_carotene.html">beta_carotene</a></code>

        </dt>
        <dd>beta_carotene</dd>
      </dl><dl><dt>

          <code><a href="kasp_dat.html">kasp_dat</a></code>

        </dt>
        <dd>kasp_dat</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

