<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Make KASP marker genotyping QC plot. — kasp_qc_ggplot • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make KASP marker genotyping QC plot. — kasp_qc_ggplot"><meta name="description" content="Make KASP marker genotyping QC plot."><meta property="og:description" content="Make KASP marker genotyping QC plot."><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Make KASP marker genotyping QC plot.</h1>
      <small class="dont-index">Source: <a href="https://github.com/awkena/panGenomeBreedr/blob/HEAD/R/kasp_validation.R" class="external-link"><code>R/kasp_validation.R</code></a></small>
      <div class="d-none name"><code>kasp_qc_ggplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Make KASP marker genotyping QC plot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">kasp_qc_ggplot</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  FAM <span class="op">=</span> <span class="st">"X"</span>,</span>
<span>  HEX <span class="op">=</span> <span class="st">"Y"</span>,</span>
<span>  geno_call <span class="op">=</span> <span class="st">"Call"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"Color"</span>,</span>
<span>  snp_id <span class="op">=</span> <span class="st">"SNPID"</span>,</span>
<span>  blank <span class="op">=</span> <span class="st">"NTC"</span>,</span>
<span>  Group_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pdf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"kasp_qc"</span>,</span>
<span>  expand_axis <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  legend.pos.x <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  legend.pos.y <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  legend.box <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>  legend.pos <span class="op">=</span> <span class="st">"inside"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  text_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A list object of KASP genotype calls processed by the `kasp_color()`
function.</p></dd>


<dt id="arg-fam">FAM<a class="anchor" aria-label="anchor" href="#arg-fam"></a></dt>
<dd><p>A character indicating the column name for FAM fluorescence
coordinates in <code>x</code>.</p></dd>


<dt id="arg-hex">HEX<a class="anchor" aria-label="anchor" href="#arg-hex"></a></dt>
<dd><p>A character indicating the column name for HEX fluorescence
coordinates in <code>x</code>.</p></dd>


<dt id="arg-geno-call">geno_call<a class="anchor" aria-label="anchor" href="#arg-geno-call"></a></dt>
<dd><p>A character value indicating the column name for KASP genotype
calls in <code>x</code>.</p></dd>


<dt id="arg-color">color<a class="anchor" aria-label="anchor" href="#arg-color"></a></dt>
<dd><p>A character value indicating the column name for assigned colors
in <code>x</code>.</p></dd>


<dt id="arg-snp-id">snp_id<a class="anchor" aria-label="anchor" href="#arg-snp-id"></a></dt>
<dd><p>A character value indicating the column name for SNP IDs
in <code>x</code>.</p></dd>


<dt id="arg-blank">blank<a class="anchor" aria-label="anchor" href="#arg-blank"></a></dt>
<dd><p>A character value indicating `No Template Controls (NTC)`
genotype calls.</p></dd>


<dt id="arg-group-id">Group_id<a class="anchor" aria-label="anchor" href="#arg-group-id"></a></dt>
<dd><p>A character value for the column ID indicating the predictions
of the positive controls in <code>x</code>.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>A logical value indicating whether to scale FAM and HEX axis
to values between 0 and 1.</p></dd>


<dt id="arg-pdf">pdf<a class="anchor" aria-label="anchor" href="#arg-pdf"></a></dt>
<dd><p>A logical value indicating whether to save plot as a pdf graphic
device when TRUE or output plot in R when FALSE.</p></dd>


<dt id="arg-width">width<a class="anchor" aria-label="anchor" href="#arg-width"></a></dt>
<dd><p>A numeric value for the width of pdf device.</p></dd>


<dt id="arg-height">height<a class="anchor" aria-label="anchor" href="#arg-height"></a></dt>
<dd><p>A numeric value for the height of pdf device.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>A character value for path or file name for saving pdf.</p></dd>


<dt id="arg-expand-axis">expand_axis<a class="anchor" aria-label="anchor" href="#arg-expand-axis"></a></dt>
<dd><p>A numeric value to expand the axes for legend positioning.</p></dd>


<dt id="arg-legend-pos-x">legend.pos.x<a class="anchor" aria-label="anchor" href="#arg-legend-pos-x"></a></dt>
<dd><p>A numeric value representing the x coordinate for legend
placement.</p></dd>


<dt id="arg-legend-pos-y">legend.pos.y<a class="anchor" aria-label="anchor" href="#arg-legend-pos-y"></a></dt>
<dd><p>A numeric value representing the y coordinate for legend
placement.</p></dd>


<dt id="arg-legend-box">legend.box<a class="anchor" aria-label="anchor" href="#arg-legend-box"></a></dt>
<dd><p>A character value of either `horizontal` or `vertical`
legend placement.</p></dd>


<dt id="arg-legend-pos">legend.pos<a class="anchor" aria-label="anchor" href="#arg-legend-pos"></a></dt>
<dd><p>A character value for the position of legend; the
default value is `inside`.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>A numeric value between 0 and 1 for modifying the
opacity of colors.</p></dd>


<dt id="arg-text-size">text_size<a class="anchor" aria-label="anchor" href="#arg-text-size"></a></dt>
<dd><p>A numeric value for setting text size.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other valid arguments that can be passed to ggplot2.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A graphic object or plot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://awkena.github.io/panGenomeBreedr">panGenomeBreedr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Assign KASP colors to plates</span></span></span>
<span class="r-in"><span><span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="kasp_color.html">kasp_color</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">panGenomeBreedr</span><span class="fu">::</span><span class="va"><a href="kasp_dat.html">kasp_dat</a></span>,</span></span>
<span class="r-in"><span>                   subset <span class="op">=</span> <span class="st">'MasterPlate'</span>,</span></span>
<span class="r-in"><span>                   sep <span class="op">=</span> <span class="st">':'</span>,</span></span>
<span class="r-in"><span>                   geno_call <span class="op">=</span> <span class="st">'Call'</span>,</span></span>
<span class="r-in"><span>                   uncallable <span class="op">=</span> <span class="st">'Uncallable'</span>,</span></span>
<span class="r-in"><span>                   unused <span class="op">=</span> <span class="st">'?'</span>,</span></span>
<span class="r-in"><span>                   blank <span class="op">=</span> <span class="st">'NTC'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0392_P01_d2 failed! Check genotype calls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0392_P01_d1 failed! Check genotype calls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0395_P01_d2 failed! Check genotype calls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0395_P01_d1 failed! Check genotype calls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0397_P01_d2 failed! Check genotype calls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Marker in Plate SE-24-0397_P01_d1 failed! Check genotype calls.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># KASP QC plot for Plate 12</span></span></span>
<span class="r-in"><span><span class="fu">kasp_qc_ggplot</span><span class="op">(</span>x <span class="op">=</span> <span class="va">dat1</span><span class="op">[</span><span class="fl">12</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                    pdf <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                    Group_id <span class="op">=</span> <span class="st">'Group'</span>,</span></span>
<span class="r-in"><span>                    scale <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                    expand_axis <span class="op">=</span> <span class="fl">0.6</span>,</span></span>
<span class="r-in"><span>                    alpha <span class="op">=</span> <span class="fl">0.5</span>,</span></span>
<span class="r-in"><span>                    legend.pos.x <span class="op">=</span> <span class="fl">0.6</span>,</span></span>
<span class="r-in"><span>                    legend.pos.y <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`SE-24-0396_P01_d1`</span>
<span class="r-plt img"><img src="kasp_qc_ggplot-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

