<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Design KASP markers based on causal variants. — kasp_marker_design • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design KASP markers based on causal variants. — kasp_marker_design"><meta name="description" content="Design KASP markers based on causal variants."><meta property="og:description" content="Design KASP markers based on causal variants."><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Design KASP markers based on causal variants.</h1>
      <small class="dont-index">Source: <a href="https://github.com/awkena/panGenomeBreedr/blob/HEAD/R/kasp_marker_design.R" class="external-link"><code>R/kasp_marker_design.R</code></a></small>
      <div class="d-none name"><code>kasp_marker_design.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Design KASP markers based on causal variants.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">kasp_marker_design</span><span class="op">(</span></span>
<span>  vcf_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gt_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="st">"variant_id"</span>,</span>
<span>  chrom_col <span class="op">=</span> <span class="st">"chrom"</span>,</span>
<span>  pos_col <span class="op">=</span> <span class="st">"pos"</span>,</span>
<span>  ref_al_col <span class="op">=</span> <span class="st">"ref"</span>,</span>
<span>  alt_al_col <span class="op">=</span> <span class="st">"alt"</span>,</span>
<span>  geno_start <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  <span class="va">marker_ID</span>,</span>
<span>  chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">genome_file</span>,</span>
<span>  save_alignment <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  region_name <span class="op">=</span> <span class="st">"loc_1"</span>,</span>
<span>  maf <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-vcf-file">vcf_file<a class="anchor" aria-label="anchor" href="#arg-vcf-file"></a></dt>
<dd><p>Path to the vcf file containing identified variants.</p></dd>


<dt id="arg-gt-df">gt_df<a class="anchor" aria-label="anchor" href="#arg-gt-df"></a></dt>
<dd><p>A data frame or matrix containing the meta data of identified
variants and sample VCF genotype calls. The variants are rows and samples as
columns.</p></dd>


<dt id="arg-variant-id-col-chrom-col-pos-col">variant_id_col, chrom_col, pos_col<a class="anchor" aria-label="anchor" href="#arg-variant-id-col-chrom-col-pos-col"></a></dt>
<dd><p>A character value specifying the
column names of variant IDs, chromosome, and positions in `gt_df` or `vcf_file`.</p></dd>


<dt id="arg-ref-al-col-alt-al-col-">ref_al_col, alt_al_col, <a class="anchor" aria-label="anchor" href="#arg-ref-al-col-alt-al-col-"></a></dt>
<dd><p>A character value specifying the column names
of reference and alternate alleles, respectively in `gt_df` or `vcf_file`.</p></dd>


<dt id="arg-geno-start">geno_start<a class="anchor" aria-label="anchor" href="#arg-geno-start"></a></dt>
<dd><p>An integer value specifying the column index number of the
start of the sample genotypes in `gt_df` or `vcf_file`.</p></dd>


<dt id="arg-marker-id">marker_ID<a class="anchor" aria-label="anchor" href="#arg-marker-id"></a></dt>
<dd><p>Designated name of variant for marker design. Name must be
contained in `gt_df` or `vcf_file`.</p></dd>


<dt id="arg-chr">chr<a class="anchor" aria-label="anchor" href="#arg-chr"></a></dt>
<dd><p>A character value representation the chromosome description in the
`genome_file`. Providing this helps to save memory in R.</p></dd>


<dt id="arg-genome-file">genome_file<a class="anchor" aria-label="anchor" href="#arg-genome-file"></a></dt>
<dd><p>Path to reference genome file in fasta format, either
compressed (.gz) or decompressed.</p></dd>


<dt id="arg-save-alignment">save_alignment<a class="anchor" aria-label="anchor" href="#arg-save-alignment"></a></dt>
<dd><p>A logical value indicating whether to save the alignment
plot to a directory specified by `plot_file`.</p></dd>


<dt id="arg-plot-file">plot_file<a class="anchor" aria-label="anchor" href="#arg-plot-file"></a></dt>
<dd><p>Path to save the sequence alignment if `plot_draw = TRUE`.</p></dd>


<dt id="arg-region-name">region_name<a class="anchor" aria-label="anchor" href="#arg-region-name"></a></dt>
<dd><p>A n optional character value for assigned region name.</p></dd>


<dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>A numeric value between 0 and 1 representing the minor allele
frequency for variant subset.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list object of two components comprising a data frame containing
all information required for KASP marker design, and a plot of DNA sequence
alignment to the reference genome.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function provides the intertek sequence to be used for marker development
for the selected casual variants. It provides all the information of allele and
location to fill Intertek form.It needs a vcf file of variants calls, and a genome
sequence of the target crop in fasta format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Example to design a KASP marker on a substitution variant</span></span></span>
<span class="r-in"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">path1</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/awkena/panGB/main/Chr02.fa.gz"</span></span></span>
<span class="r-in"><span><span class="va">path2</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Sobic.002G302700_SNP_snpeff.vcf"</span>,</span></span>
<span class="r-in"><span>                     package <span class="op">=</span> <span class="st">"panGenomeBreedr"</span>,</span></span>
<span class="r-in"><span>                     mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ma1</span> <span class="op">&lt;-</span> <span class="fu">kasp_marker_design</span><span class="op">(</span>vcf_file <span class="op">=</span> <span class="va">path2</span>,</span></span>
<span class="r-in"><span>                          variant_id_col <span class="op">=</span> <span class="st">'ID'</span>,</span></span>
<span class="r-in"><span>                          chrom_col <span class="op">=</span> <span class="st">'CHROM'</span>,</span></span>
<span class="r-in"><span>                          pos_col <span class="op">=</span> <span class="st">'POS'</span>,</span></span>
<span class="r-in"><span>                          ref_al_col <span class="op">=</span> <span class="st">'REF'</span>,</span></span>
<span class="r-in"><span>                          alt_al_col <span class="op">=</span> <span class="st">'ALT'</span>,</span></span>
<span class="r-in"><span>                          genome_file <span class="op">=</span> <span class="va">path1</span>,</span></span>
<span class="r-in"><span>                          geno_start <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                          marker_ID <span class="op">=</span> <span class="st">"SNP_Chr02_69200443"</span>,</span></span>
<span class="r-in"><span>                          chr <span class="op">=</span> <span class="st">"Chr02"</span>,</span></span>
<span class="r-in"><span>                          save_alignment <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                          plot_file <span class="op">=</span> <span class="va">path</span>,</span></span>
<span class="r-in"><span>                          region_name <span class="op">=</span> <span class="st">"ma1"</span>,</span></span>
<span class="r-in"><span>                          maf <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> using Gonnet</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># View marker alignment output from temp folder</span></span></span>
<span class="r-in"><span><span class="co"># path3 &lt;- file.path(path, list.files(path = path, "alignment_"))</span></span></span>
<span class="r-in"><span><span class="co"># system(paste0('open "', path3, '"'))</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/on.exit.html" class="external-link">on.exit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

